source/extensions/quic_listeners/quiche/codec_impl.cc

```diff
@@ -1,41 +1,93 @@
 namespace Quic {
 
 bool QuicHttpConnectionImplBase::wantsToWrite() { return quic_session_.HasDataToWrite(); }
 
+void QuicHttpConnectionImplBase::runWatermarkCallbacksForEachStream(
+    quic::QuicSmallMap<quic::QuicStreamId, std::unique_ptr<quic::QuicStream>, 10>& stream_map,
+    bool high_watermark) {
+  for (auto& it : stream_map) {
+    if (!it.second->is_static()) {
+      auto stream = dynamic_cast<EnvoyQuicStream*>(it.second.get());
+      if (high_watermark) {
+        // Only call watermark callbacks on non QUIC static streams which are
+        // crypto stream and Google QUIC headers stream.
+        ENVOY_LOG(debug, "runHighWatermarkCallbacks on stream {}", it.first);
+        stream->runHighWatermarkCallbacks();
+      } else {
+        ENVOY_LOG(debug, "runLowWatermarkCallbacks on stream {}", it.first);
+        stream->runLowWatermarkCallbacks();
+      }
+    }
+  }
+}
+
 QuicHttpServerConnectionImpl::QuicHttpServerConnectionImpl(
     EnvoyQuicServerSession& quic_session, Http::ServerConnectionCallbacks& callbacks)
     : QuicHttpConnectionImplBase(quic_session), quic_server_session_(quic_session) {
   quic_session.setHttpConnectionCallbacks(callbacks);
 }
 
 void QuicHttpServerConnectionImpl::onUnderlyingConnectionAboveWriteBufferHighWatermark() {
-  for (auto& it : quic_server_session_.stream_map()) {
-    if (!it.second->is_static()) {
-      // Only call watermark callbacks on non QUIC static streams which are
-      // crypto stream and Google QUIC headers stream.
-      ENVOY_LOG(debug, "runHighWatermarkCallbacks on stream {}", it.first);
-      dynamic_cast<EnvoyQuicServerStream*>(it.second.get())->runHighWatermarkCallbacks();
-    }
-  }
+  runWatermarkCallbacksForEachStream(quic_server_session_.stream_map(), true);
 }
 
 void QuicHttpServerConnectionImpl::onUnderlyingConnectionBelowWriteBufferLowWatermark() {
-  for (const auto& it : quic_server_session_.stream_map()) {
-    if (!it.second->is_static()) {
-      ENVOY_LOG(debug, "runLowWatermarkCallbacks on stream {}", it.first);
-      dynamic_cast<EnvoyQuicServerStream*>(it.second.get())->runLowWatermarkCallbacks();
-    }
-  }
+  runWatermarkCallbacksForEachStream(quic_server_session_.stream_map(), false);
 }
 
 void QuicHttpServerConnectionImpl::goAway() {
   quic_server_session_.SendGoAway(quic::QUIC_PEER_GOING_AWAY, "server shutdown imminent");
 }
 
+QuicHttpClientConnectionImpl::QuicHttpClientConnectionImpl(EnvoyQuicClientSession& session,
+                                                           Http::ConnectionCallbacks& callbacks)
+    : QuicHttpConnectionImplBase(session), quic_client_session_(session) {
+  session.setHttpConnectionCallbacks(callbacks);
+}
+
+Http::StreamEncoder&
+QuicHttpClientConnectionImpl::newStream(Http::StreamDecoder& response_decoder) {
+  auto stream = dynamic_cast<EnvoyQuicClientStream*>(
+      quic_client_session_.CreateOutgoingBidirectionalStream());
+  ASSERT(stream != nullptr, "Fail to create QUIC stream.");
```

Comment:
nit: I would remove this assert and ones like it. It should fail in a very obvious way on the next line.

Reply:
With HTTP/2 if we overflow max streams nghttp2 will fail the stream and it will "work." So yeah we need to TODO this. I actually though this assert was checking the dynamic cast or something else (one of the reasons I don't like dynamic casts if they can be avoided).

Reply:
What do you mean by it will "work"?
nghttp2_submit_request() returns NGHTTP2_ERR_STREAM_ID_NOT_AVAILABLE upon exceeding stream limit. But I saw envoy assumes it always return a valid stream id: https://github.com/envoyproxy/envoy/blob/48c7c203890cf6c011fa6591a3b233d637c334f5/source/common/http/http2/codec_impl.cc#L274.

Reply:
Hmm I haven't looked at this in a long time. I thought we handled this case somehow but I could be wrong. If we don't handle this case we should definitely do so. Can we open a separate issue on this? cc @alyssawilk.